### PR TITLE
feat(log): track request id in the log message

### PR DESCRIPTION
**Pull request by heww**
Track request id in the log message.

Signed-off-by: He Weiwei <hweiwei@vmware.com>

### CODE REVIEW

File: src/common/utils/log/logger.go [hunk 1]
```diff
@@ -66,11 +76,20 @@ func New(out io.Writer, fmtter Formatter, lvl Level, options ...interface{}) *Lo
 			depth = d
 		}
 	}
+
+	mu, ok := writerMutexMap[out]
```
Comment (ywk253100) [inline src/common/utils/log/logger.go]: What is this used for?

File: src/common/utils/log/logger.go [hunk 2]
```diff
@@ -79,6 +98,58 @@ func DefaultLogger() *Logger {
...
+		}
+		for key, value := range fields {
+			copyFields[key] = value
+		}
+
+		sortedKeys := make([]string, 0, len(copyFields))
+		for key := range copyFields {
+			sortedKeys = append(sortedKeys, key)
+		}
+		sort.Strings(sortedKeys)
```
Comment (ywk253100) [inline src/common/utils/log/logger.go]: After sorting the fields will not be the same sequence with the original one, is this expected?
FuncC(with field c)->funcB(with filed b)->FuncA(with filed a), the log will be: [a,b,c]xxx, right?

Comment (heww) [reply src/common/utils/log/logger.go]: 1. In Golang, iterations over maps are random, sorting here can make the fileds output is ordered in log

2. `FuncC(with field c)->funcB(with filed b)->FuncA(with filed a), the log will be: [a,b,c]xxx, right?` No, the log will be [c]xxx in `FuncC`, [b]xxx in `funcB` and [a]xxx in `FuncA`. `log.G(ctx)` returns the logger saved in the ctx (if not found, returns the default one), and `logger.WithFields` does not modify the logger itself, it returns new logger.

File: src/common/utils/log/logger.go (hunk 1 above)
Comment (heww) [reply src/common/utils/log/logger.go]: This is used to lock the output writer, when multi logger use the same output writer.

En, this is may not needed for the is a root logger in the system (the default logger), and other loggers are clone from this. I will remove the `writerMutexMap`.

File: src/common/utils/log/logger.go [hunk 3]
```diff
@@ -79,6 +88,58 @@ func DefaultLogger() *Logger {
...
+// WithDepth ...
+func (l *Logger) WithDepth(depth int) *Logger {
+	r := l.clone()
+	r.callDepth = depth
+
+	return r
+}
+
+// WithFields ...
+func (l *Logger) WithFields(fields Fields) *Logger {
```
Comment (reasonerjt) [inline src/common/utils/log/logger.go]: Can we make Withxx and Setxxx consistent?  Or protect Setxxx from being called by mistake?

File: src/server/middleware/log/log.go [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+	"github.com/goharbor/harbor/src/common/utils/log"
+	"github.com/goharbor/harbor/src/server/middleware"
+)
+
+// Middleware middleware which add logger to context
+func Middleware() func(http.Handler) http.Handler {
+	return middleware.New(func(w http.ResponseWriter, r *http.Request, next http.Handler) {
+		rid := r.Header.Get("X-Request-ID")
+		if rid != "" {
+			ctx := log.WithLogger(r.Context(), log.L.WithFields(log.Fields{"requestID": rid}))
```
Comment (reasonerjt) [inline src/server/middleware/log/log.go]: Would be helpful for debugging if we can log a message with request id and basic information of the request?

Comment (heww) [reply src/server/middleware/log/log.go]: debug info added

File: src/common/utils/log/logger.go (hunk 3 above)
Comment (heww) [reply src/common/utils/log/logger.go]: SetXxx changed to private method to make logger is immutable.

**Pull request merged**